src/holder.rs

```diff
@@ -70,38 +115,95 @@ impl<'domain, F> HazardPointer<'domain, F> {
         'o: 'l,
         F: 'static,
     {
-        self.hazard.protect(ptr as *mut u8);
-
-        crate::asymmetric_light_barrier();
-
-        let ptr2 = src.load(Ordering::Acquire);
-        if ptr != ptr2 {
-            self.hazard.reset();
-            Err(ptr2)
-        } else {
-            // All good -- protected
-            Ok(std::ptr::NonNull::new(ptr).map(|nn| {
-                // Safety: this is safe because:
-                //
-                //  1. Target of ptr1 will not be deallocated for the returned lifetime since
-                //     our hazard pointer is active and pointing at ptr1.
-                //  2. Pointer address is valid by the safety contract of load.
-                let r = unsafe { nn.as_ref() };
-                debug_assert_eq!(
-                    self.domain as *const Domain<F>,
-                    r.domain() as *const Domain<F>,
-                    "object guarded by different domain than holder used to access it"
-                );
-                r
-            }))
-        }
+        unsafe { try_protect(&self.domain, &self.hazard, ptr, src) }
     }
 
     pub fn reset_protection(&mut self) {
         self.hazard.reset();
     }
 }
 
+///
```

Comment:
It's hard to spot at a glance — were there any changes to the bodies of these functions?

Reply:
No the only change was to pull them out of the `self` context so I could re-use them between the Many and the Single variants